conwayste/src/ui/checkbox.rs

```diff
@@ -0,0 +1,175 @@
+
+/*  Copyright 2019 the Conwayste Developers.
+ *
+ *  This file is part of conwayste.
+ *
+ *  conwayste is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  conwayste is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with conwayste.  If not, see
+ *  <http://www.gnu.org/licenses/>. */
+use chromatica::css;
+
+use ggez::graphics::{self, Rect, Font, Color, DrawMode, DrawParam};
+use ggez::nalgebra::{Point2, Vector2};
+use ggez::{Context, GameResult};
+
+use std::rc::Rc;
+
+use super::{
+    label::Label,
+    widget::Widget,
+    helpe::within_widget,
+    UIAction,
+    WidgetID,
+};
+
+// PR_GATE clean me up scotty (per PR feedback)
+#[derive(PartialEq, Clone, Copy, Debug)]
+pub enum ToggleState {
+    Disabled,
+    Enabled
+}
+
+pub struct Checkbox {
+    pub id:     WidgetID,
+    pub label: Label,
+    pub state: ToggleState,
+    pub dimensions: Rect,
+    pub hover: bool,
+    pub action: UIAction
+}
+
+const LABEL_OFFSET_X: f32 = 30.0;
+const LABEL_OFFSET_Y: f32 = -5.0;
+
+/// A standard checkbox widget that can be enabled or disabled via the ToggleState structure.
+impl Checkbox {
+    /// Creates a Checkbox widget.
+    ///
+    /// # Arguments
+    /// * `ctx` - GGEZ context
+    /// * `widget_id` - Unique widget identifier
+    /// * `action` - Unique action identifer
```

Comment:
What happens if this is not unique?

Reply:
If it's not unique, you'll end up interacting with the wrong widget, likely leading to undesired or buggy behavior in the application.

Reply:
Doesn't the `widget_id` prevent that?

Reply:
It did back when `WidgetID` was an enum. Now it doesn't since its just a usize value, which a user could accidentally re-use. We could keep track of what ID's are in use and panic if they try to re-use an ID during widget creation. LMK what you think.